Run compatibility tests in the real browsers (closes #2)

just-boris opened the pull request:
Added test suite, which runs the same test fixtures in real browsers and compares results.

Karma is not necessary to do it. It can be done using `eval` method in the WebDriver API, which executes a script in remote browser.

To run tests in SauceLabs you need a key. Provide key and username to tests via environment variables:
```
SAUCE_USER=<user> SAUCE_KEY=<key> npm test
```

jeffcarp commented:
Awesome, thank you @just-boris! I'll try adding Sauce credentials to Travis and see if I can get it to run.

It looks like this code is designed to work with io.js/Node >= v3 (looking at use of `const`). I'm okay with requiring Node v4 in order to run the tests (the .travis.yml file will have to be updated, though).

just-boris commented:
Updated `.travis.yml`.
Now build is passing fine

jeffcarp commented on test/browser.js:
```diff
@@ -0,0 +1,69 @@
+
+const browsers = ['chrome', 'firefox', 'internet explorer'];
+
+if (process.env.SAUCE_USER) {
```
It looks like Sauce uses `SAUCE_USERNAME` instead of `SAUCE_USER`. I think it might be more straightforward if we made this match the Sauce lingo.

jeffcarp commented on test/browser.js:
```diff
@@ -0,0 +1,69 @@
+const assert = require('assert');
+const wd = require('wd');
+const util = require('./util');
+const abab = require('..').btoa;
```
This needs to be `const abab = require('..');`

jeffcarp commented on test/browser.js:
```diff
@@ -0,0 +1,69 @@
+            const expectedDescriptor = util.stripChars(expected);
+
+            it(`correctly converts ${inputDescriptor} into ${expectedDescriptor}`, function () {
+              return browser.eval(`atob('${input}')`).then(function (result) {
```
I think the `eval`'d statement needs to be changed - for cases that don't throw a `TypeError`, it works, but for cases when it does, it fails.

jeffcarp replied:
I tried writing a try/catch in the `eval`'d string but couldn't get it to work. I'll try again later.

just-boris commented:
Fixed codestyle issues.

To catch exception and return null instead we need wrap code in function expression. I did it and now more tests are passing. Also I got issues regarding character escape and fixed them using `JSON.stringify()`.

just-boris commented:
Also note that we cannot test  `atob({toString: function() {return 'foo';}})` by this way, because it serialized when script is generated.

Maybe we need to create special set of test cases against browsers

jeffcarp commented:
Awesome, thank you @just-boris :smile:

Interesting re: the `toString` cases. Do you think it would be a good solution to add a separate set of cases which are strings to be eval'd? (e.g. `const evalCases = ['{toString: function() {return 'foo';}}'];`) I feel like that way all cases would work in all environments.

jeffcarp commented:
Also I really like the ability to skip the browser tests in development, but I think CI should always run them. Interested to hear your thoughts on this.

just-boris commented:
I'll try to read fixtures file as text and pass entire file to the browser. Now it exports data via `module.exports`. I can process file with browserify to make it work.

----

This solution got a lot of overhead. Maybe karma can help to make it simpler. I'll try to do it too

just-boris commented:
Browser test are run when Sauce credentials presents in environment. Travis [can store private data](http://docs.travis-ci.com/user/encryption-keys/) like access key in the `.travis.yml` in encrypted form. Also this data available only for master build and hidden on builds from forks, so you will get exactly what you want. [This repo](https://github.com/JamesMGreene/document.currentScript/blob/master/.travis.yml) already has an example of this setup

jeffcarp commented:
Closing in favor of #8 - please re-open if I missed something
